swap/src/bitcoin.rs

```diff
@@ -249,6 +249,61 @@ pub fn current_epoch(
     ExpiredTimelocks::None
 }
 
+/// Bitcoin error codes: https://github.com/bitcoin/bitcoin/blob/97d3500601c1d28642347d014a6de1e38f53ae4e/src/rpc/protocol.h#L23
+pub enum RpcErrorCode {
+    /// Transaction or block was rejected by network rules. Error code -26.
+    RpcVerifyRejected,
+    /// Transaction or block was rejected by network rules. Error code -27.
+    RpcVerifyAlreadyInChain,
+    /// General error during transaction or block submission
+    RpcVerifyError,
+}
+
+impl From<RpcErrorCode> for i64 {
+    fn from(code: RpcErrorCode) -> Self {
+        match code {
+            RpcErrorCode::RpcVerifyError => -25,
+            RpcErrorCode::RpcVerifyRejected => -26,
+            RpcErrorCode::RpcVerifyAlreadyInChain => -27,
+        }
+    }
+}
+
+pub fn ok_if_rpc_error_code(
+    error: &anyhow::Error,
+    expected_error_code: RpcErrorCode,
+) -> anyhow::Result<()> {
+    let string = match error.downcast_ref::<bdk::Error>() {
+        Some(bdk::Error::Electrum(bdk::electrum_client::Error::Protocol(
+            serde_json::Value::String(string),
+        ))) => string,
+        _ => bail!("Error is of incorrect variant:{}", error),
+    };
+
+    let json = serde_json::from_str(&string.replace("sendrawtransaction RPC error:", ""))?;
+
+    let json_map = match json {
+        serde_json::Value::Object(map) => map,
+        _ => bail!("Json error is not json object "),
+    };
+
+    let error_code_value = match json_map.get("code") {
+        Some(val) => val,
+        None => bail!("No error code field"),
+    };
+
+    let error_code_number = match error_code_value {
+        serde_json::Value::Number(num) => num,
+        _ => bail!("Error code is not a number"),
+    };
+
+    if *error_code_number != serde_json::Number::from(i64::from(expected_error_code)) {
+        bail!("Unexpected error code: {}", error_code_number)
+    }
```

Comment:
Minor nit: I think it would be slightly cleaner to just return the parsed error code from this function and allow the caller to decide what they want to do with it. Gives you a bit more flexibility in how it can be used.

Instead of:

```rust
if ok_if_rpc_error_code(&err, RpcErrorCode::RpcVerifyAlreadyInChain).is_ok() {

}
```

You would say:

```rust
if parse_rpc_error_code(&err)? == RpcErrorCode::RpcVerifyAlreadyInChain {

}
```